Issue #5 Allow public:// prefix for local unmanaged files

dsenalik opened the pull request:
This partially addresses issue #5 and issue #4 by allowing files already on the filesystem, but not managed by Tripal, to be linked using a public:// prefix.
This allows the files to have size and md5 checksum values populated.
For example I tested with both a URL and a file in an existing heirarchy
```
https://static-content.springer.com/esm/art%3A10.1186%2F1471-2164-12-389/MediaObjects/12864_2011_3596_MOESM2_ESM.FASTA
public://bulk_data/carrotomics/sequence/est-assembly1/12864_2011_3596_MOESM2_ESM.FASTA.gz
```
![20210310_downloadexample](https://user-images.githubusercontent.com/8419404/110712535-afed3c00-81c6-11eb-8375-5df1a5ec2d14.png)

spficklin commented on includes/TripalFields/schema__itemlocation/schema__itemlocation_widget.inc:
```diff
@@ -75,6 +75,10 @@ class schema__itemlocation_widget extends ChadoFieldWidget {
         $query->condition('FM.uri', $uri);
         $local = $query->execute()->fetchField();
+        // we can also use the public:// prefix with local files not managed by tripal
+        if (!$local) {
```
Rather then check if $local is empty here, how about adding another if statement on line 71:

```
if (preg_match('/^public:\/\//', $uri) {
  remote = $uri
}
```

dsenalik replied:
I didn't do it that way because tripal managed files can also have a public:// prefix, e.g. ```public://tripal/users/1/T119314.tree```
As best as I could figure, I need to see if it is managed first.

spficklin replied:
I see. Okay,  I think that bit of information gets lost then in the code, and I think the code makes an assumption that anything that isn't an http:// or ftp:// is local.  That wasn't your fault, it was there before.  So, to help clear that up and make the condition you raise more clear how about this change:

```php
if ($uri) {
  // Is the URI provided by the user a managed Drupal file?
  $query = db_select('file_managed', 'FM');
  $query->fields('FM', ['fid']);
  $query->distinct();
  $query->condition('FM.uri', $uri);
  $local = $query->execute()->fetchField();

  // If the file is not local, we will consider it a remote file. That should include a file that is on a different server
  // (eg. via ftp:// or http://) or local to the server, in the Drupal files folder, but not managed by Drupal (e.g. public://).
  // The latter case would occur if the site admin manually added files to the Drupal files folder and didn't use a 
  // Drupal or Tripal uploader.
  if (!$local) {
    $remote = $uri;
  }
}
```
That code could replace lines 68-83

dsenalik replied:
There are new changes, see what you think

spficklin commented:
This is a nice addition @dsenalik thanks for submitting.   I just have one code change request that's simple and noted above.

dsenalik commented:
I've had some thoughts about reworking this a bit to be more robust. watch for an update

dsenalik commented:
Enhanced validation and I renamed "local" to "managed" and "remote" to "unmanaged" to hopefully make things more clear, since remote included local files when not managed by tripal. I hope this doesn't break anything.
I also came across two additional bugs and fixed them
1. files with ftp:// prefix lost their url because they were eaten by file_create_url(), this change is in schema_itemlocation.inc
2. my site name wasn't substituted because it is using https instead of http, this change is in schema_itemlocation_formatter.inc

spficklin commented:
This change looks good. I think we can commit and if anything else arises that we didn't see we can fix later. Thanks for all the fixes!
